### PR TITLE
feature(planner): Support subqueries in new planner

### PR DESCRIPTION
**Pull request by leiysky**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Support planning subqueries. The execution part will be done in other PRs.

## Changelog

- New Feature

## Related Issues

Part of #5210

### CODE REVIEW

File: query/src/sql/exec/expression_builder.rs [hunk 1]
```diff
@@ -178,7 +180,7 @@ impl<'a> ExpressionBuilder<'a> {
                         .iter()
                         .map(|arg| self.normalize_aggr_to_col(arg.clone()))
                         .collect::<Result<Vec<Expression>>>()?,
-                })
+                });
             }
             _ => {}
```
Comment (zhang2014) [inline query/src/sql/exec/expression_builder.rs:185]:
```suggestion
            _ => Err(ErrorCode::LogicaError("It's a bug, epxression type missing match"))
```

File: query/src/sql/planner/binder/aggregate.rs [hunk 2]
```diff
@@ -62,16 +61,16 @@ impl Binder {
...
         group_by_expr: &[Expr],
         input_context: &mut BindContext,
     ) -> Result<()> {
-        let scalar_binder = ScalarBinder::new(input_context);
-        let group_expr = group_by_expr
-            .iter()
-            .map(|expr| scalar_binder.bind_expr(expr))
-            .collect::<Result<Vec<(Scalar, DataTypeImpl)>>>()?;
+        let scalar_binder = ScalarBinder::new(input_context, self.ctx.clone());
+        let mut group_expr = vec![];
```
Comment (zhang2014) [inline query/src/sql/planner/binder/aggregate.rs:70]: Vec::with_capacity() is better.

File: query/src/sql/planner/binder/aggregate.rs [hunk 3]
```diff
@@ -62,16 +61,16 @@ impl Binder {
         Ok(())
     }
 
-    pub(super) fn bind_group_by(
+    pub(super) async fn bind_group_by(
```
[the remark this answers is not in the document]
Comment (leiysky) [reply query/src/sql/planner/binder/aggregate.rs:64]: We may process subquery, which may access `Catalog`.

I think we need synchronous `Catalog` API.

File: query/src/sql/planner/binder/sort.rs [hunk 4]
```diff
@@ -13,28 +13,26 @@
...
     ) -> Result<()> {
-        let scalar_binder = ScalarBinder::new(bind_context);
-        let order_by_exprs = order_by
-            .iter()
-            .map(|expr| scalar_binder.bind_expr(&expr.expr))
-            .collect::<Result<Vec<(Scalar, DataTypeImpl)>>>()?;
+        let scalar_binder = ScalarBinder::new(bind_context, self.ctx.clone());
+        let mut order_by_exprs = vec![];
+        for expr in order_by {
+            order_by_exprs.push(scalar_binder.bind_expr(&expr.expr).await?);
```
Comment (andylokandy) [inline query/src/sql/planner/binder/sort.rs:34]: You may consider using https://docs.rs/futures/latest/futures/prelude/stream/struct.FuturesOrdered.html

File: query/src/sql/planner/binder/scalar.rs [hunk 5]
```diff
@@ -12,26 +12,30 @@
...
-    pub fn new(bind_context: &'a BindContext) -> Self {
-        ScalarBinder { bind_context }
+    pub fn new(bind_context: &'a BindContext, ctx: Arc<QueryContext>) -> Self {
+        ScalarBinder { bind_context, ctx }
     }
 
-    pub fn bind_expr(&self, expr: &Expr) -> Result<(Scalar, DataTypeImpl)> {
-        let type_checker = TypeChecker::new(self.bind_context);
-        type_checker.resolve(expr, None)
+    pub async fn bind_expr(&self, expr: &Expr) -> Result<(Scalar, DataTypeImpl)> {
```
Comment (sundy-li) [inline query/src/sql/planner/binder/scalar.rs:37]: I just wrote sync bind expr in #5016  ...

We really need sync API for catalog & user mgr.

**Pull request merged**